onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc

```diff
@@ -113,19 +154,29 @@ void XnnpackExecutionProvider::RegisterAllocator(AllocatorManager& allocator_man
                                                                 OrtAllocatorType::OrtDeviceAllocator));
           });
 
-      cpu_alloc = CreateAllocator(allocator_info);
+      cpu_alloc = AllocContext_global
+                      ? AllocContext_global
+                      : CreateAllocator(allocator_info);
       // enable sharing of our allocator
       allocator_manager.InsertAllocator(cpu_alloc);
     }
 
     InsertAllocator(cpu_alloc);
   }
 
-  // TODO: Create `struct xnn_allocator` that wraps cpu_allocator, and provide in the call to xnn_initialize so that
-  //       xnnpack is using the ORT allocator.
-  xnn_status st = xnn_initialize(nullptr);
-  if (st != xnn_status_success) {
-    ORT_THROW("XNNPACK initialization failed with status ", st);
+  AllocContext_use_count_g++;
+  static const xnn_allocator xnn_default_allocator{&AllocContext_global,
+                                                   xnn_allocate,
+                                                   xnn_reallocate,
+                                                   xnn_deallocate,
+                                                   xnn_aligned_allocate,
+                                                   xnn_aligned_deallocate};
+  if (!AllocContext_global) {
+    AllocContext_global = cpu_alloc;
+    xnn_status st = xnn_initialize(&xnn_default_allocator);
+    if (st != xnn_status_success) {
```

Comment:
Can we put this in the anonymous namespace and use std::call_once? that scopes all the statics and avoids doing your own counting.

```c++
void init_xnnpack(AllocatorPtr allocator) {
  static AllocatorPtr saved_ort_allocator = allocator;
  static const xnn_allocator xnn_allocator_wrapper = xnn_allocator{allocator.get(), ...};
  xnn_status st = xnn_initialize(&xnn_allocator_wrapper);
  if (st != xnn_status_success) {
    ORT_THROW("XNNPACK initialization failed with status ", st);
  }
}
```

Reply:
Good Idea!!!

Reply:
I don't think there's any need to reference count. We need a static AllocatorPtr to capture the allocator we use with xnnpack. 

I believe that essentially needs to remain alive forever given there's no way to change the allocator xnnpack uses. 

xnn_deinitializer does [nothing](https://github.com/google/XNNPACK/blob/525bf20e8fdc1a4858713899616b67d0f7977c84/src/init.c#L7920). 

xnn_initialize uses mechanisms to [only ever set the allocator once](https://github.com/google/XNNPACK/blob/525bf20e8fdc1a4858713899616b67d0f7977c84/src/init.c#L7903-L7912), regardless of how many times it's called. if we release the allocator we used with xnnpack we are dead if the user creates a new session.

Reply:
DO this with a function with a static AllocatorPtr which captures allocator. Seems it's more concise